arduino-core/src/cc/arduino/UploaderUtils.java

```diff
@@ -50,7 +50,16 @@ public Uploader getUploaderByPreferences(boolean noUploadPort) {
 
     BoardPort boardPort = null;
     if (!noUploadPort) {
-      boardPort = BaseNoGui.getDiscoveryManager().find(PreferencesData.get("serial.port"));
+      String port = PreferencesData.get("serial.port");
+      if (port == null || port.isEmpty()) {
+        return null;
+      }
+      boardPort = BaseNoGui.getDiscoveryManager().find(port);
+      //if (boardPort == null) {
+        // Is there ever a reason to attempt upload when
+        // the Port is not found by DiscoveryManager?
+        //return null;
```

Comment:
Good question. It would be interesting to investigate in what cases (if at all) `find(port)` returns `null` (I suspect it does not return `null` when an invalid serial port is given, e.g. `COM1` on Linux, since that tries to run avrdude, so perhaps this is only done with network port stuff, if at all?). It is also interesting to investigate what `newUploader` below does with a null `boardPort`.

In any case, I think a `null` `boardPort` should not return `null` here, since that shows a "Please select a port" message, which is not appropriate when the user has selected a port, but it is not available or otherwise cannot be used.

Before merging, I think at least these commented lines should be removed, even if no more elaborate handling of this case is added.

Reply:
I tried with COM1 on Linux.  Indeed boardPort is null (verified by adding a println).  Then avrdude is run.  The result is this error:
![sc](https://user-images.githubusercontent.com/965463/44909475-5857e680-acd4-11e8-91ce-0c88a439f4db.png)

Reply:
But perhaps it's best to allow the uploader to run anyway?  Maybe some boards have uploaders that can auto-detect the board, even when given a port name that doesn't exist?

Reply:
Oh, that is a bit unexpected. I would think that, since `boardPort` is `null`, the actual port name (`COM1`) is not passed into `newUploader`, so it could not pass it to `avrdude`. However, I guess the port name is read from the preferences separately when building the avrdude recipe, so the `boardPort` argument here serves a different purpose?

Reply:
Honestly, I didn't dig deeper into how it's really using boardPort.  That's the reason for the question!

But now that we're discussing this, I agree it's probably best to just allow the uploader to run even if boardPort is null.  I would worry about breaking 3rd party uploaders.

The point here was to just require first-time users to select *something* from the Ports menu.

Reply:
Commented stuff deleted.